### PR TITLE
test(colorpicker): Add tests and convert them to TL/R

### PR DESCRIPTION
**Pull request by sahlhoff**
Adds TL/R to `ColorInput` and `ColorPreview`.

#286 

It could be worth it to converge these into a single component and use the `readonly` prop for the variant `ColorPreview`. Could be a reason why we have two components but wanted to leave a note here.

### CODE REVIEW

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 1]
```diff
@@ -1,115 +1,173 @@
 import * as React from 'react';
 import ColorInput from '../lib/ColorInput';
-import {mount} from 'enzyme';
+import {render, fireEvent} from '@testing-library/react';
 import ReactDOMServer from 'react-dom/server';
 import {axe} from 'jest-axe';
```
Comment (sahlhoff) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: Going to move these tests to cypress.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 2]
```diff
@@ -1,115 +1,150 @@
...
+    });
+
+    describe('with a value', () => {
+      test('should render a ColorInput with a value', () => {
+        const {getByTestId} = render(<ColorInput value={value} data-testid={id} />);
+        expect(getByTestId(id).value).toBe(value);
+      });
+    });
+
+    describe('with a value and check', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: Not a comment on this test but I guess while you're in here with the stories, it may be good to add a story, example, or knob for the checked state, I realized that isn't demonstrated at all in our Storybook. It could also be a separate PR as well.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 3]
```diff
@@ -1,115 +1,150 @@
...
+
+    describe('when provided a value with a hashtag', () => {
+      test('the value is stripped', () => {
+        const {getByTestId} = render(<ColorInput value={'#eee'} data-testid={id} />);
+        expect(getByTestId(id).value).toBe(value);
+      });
+    });
+
+    describe('when provided more than 6 characters as the value', () => {
+      test('the value is stripped', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: This looks like a copy and paste error from above? Something like `'the value is truncated to a length of 6'` or something like that

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 4]
```diff
@@ -1,115 +1,150 @@
...
+    describe('when provided an input ref', () => {
+      test('input ref should be defined', () => {
+        const ref: React.RefObject<HTMLInputElement> = React.createRef();
+        render(<ColorInput inputRef={ref} />);
+        expect(ref.current).toBeDefined();
+      });
+    });
+
+    describe('when provided a value with a hashtag', () => {
+      test('the value is stripped', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: From a failure perspective, it may not be clear what fails. Maybe something more specific like `'the hash is stripped from the value'`.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 5]
```diff
@@ -1,115 +1,150 @@
...
+
+    describe('when provided an input ref', () => {
+      test('input ref should be defined', () => {
+        const ref: React.RefObject<HTMLInputElement> = React.createRef();
+        render(<ColorInput inputRef={ref} />);
+        expect(ref.current).toBeDefined();
+      });
+    });
+
+    describe('when provided a value with a hashtag', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: I think `#` is just 'hash' haha.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 6]
```diff
@@ -1,115 +1,150 @@
...
-describe('ColorInput Accessibility', () => {
-  test('ColorInput should pass axe DOM accessibility guidelines', async () => {
-    const html = ReactDOMServer.renderToString(
-      <FormField label="Label" inputId="input-plain">
-        <ColorInput />
-      </FormField>
-    );
-    expect(await axe(html)).toHaveNoViolations();
+  describe('when changed', () => {
+    test('should update value', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: I don't think this test is doing what you think it's doing. You cover setting the value [here](https://github.com/Workday/canvas-kit/pull/407/files#diff-8a0bed2634f532ea7599ae6a06bed58cR35).

Since this is a controlled component, we probably don't need to test for how it updates the value, that is beyond the scope of this component.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 7]
```diff
@@ -1,115 +1,150 @@
...
+      const {getByTestId} = render(<ColorInput value="" data-testid={id} />);
+
+      expect(getByTestId(id).value).toBe('');
+      getByTestId(id).value = value;
+      expect(getByTestId(id).value).toBe(value);
+    });
+
+    test('should call onChange callback on value change', () => {
```
Comment (lychyi) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: Curious what others think here, I think most of these tests in `'when changed'` could just be functional tests in Cypress. They're not really testing API contracts at this point but rather functionality of the input.

File: modules/color-picker/react/spec/ColorInput.spec.tsx (hunk 2 above)
Comment (sahlhoff) [reply modules/color-picker/react/spec/ColorInput.spec.tsx]: Yup, added a `Checked` story 😎

File: modules/color-picker/react/spec/ColorInput.spec.tsx (hunk 6 above)
Comment (sahlhoff) [reply modules/color-picker/react/spec/ColorInput.spec.tsx]: I removed this test since it was redundant.

File: cypress/integration/ColorPicker.spec.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    ['Default', 'Alert', 'Error', 'Checked', 'Grow', 'Grow with Error'].forEach(story => {
+      context(`given the '${story}' story is rendered`, () => {
+        beforeEach(() => {
+          h.stories.load(colorInputStory, story);
+        });
+
+        it('should pass accessibility checks', () => {
+          cy.checkA11y();
+        });
+      });
```
Comment (anicholls) [inline cypress/integration/ColorPicker.spec.ts]: Can you test inputting text into the input and checking if the color changes? We should be testing all interaction with the story here. Things like:

- Input a correct hex (e.g. `000000`)
- input a 3 digit hex (e.g. `000`)
- Input an invalid hex (e.g. `zzzzzz`)
- Input a string that's too long (e.g. `0000000`)
etc.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 9]
```diff
@@ -1,115 +1,140 @@
...
-  });
-  test('should set inputRef as reference to native input', () => {
-    const ref = React.createRef<HTMLInputElement>();
-    mount(<ColorInput inputRef={ref} value={''} />);
-    expect(ref.current && ref.current.tagName && ref.current.tagName.toUpperCase()).toBe('INPUT');
-  });
+  describe('when rendered', () => {
+    describe('with a placeholder', () => {
+      test('should render a ColorInput with a placeholder', () => {
+        const {getByTestId} = render(
```
Comment (anicholls) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: Based on the [recommendations in the docs](https://testing-library.com/docs/dom-testing-library/api-queries#bytestid), we should avoid `getByTestId` wherever possible. I think you'll be most successful with `getByRole('textbox')` here so it doesn't require any extra props.

File: modules/color-picker/react/spec/ColorInput.spec.tsx [hunk 10]
```diff
@@ -1,115 +1,140 @@
...
+    describe('with a value and check', () => {
+      test('should render a ColorInput with a value and check', () => {
+        const {container} = render(<ColorInput value={value} showCheck={true} />);
+
+        expect(container.querySelector('svg')).toHaveClass('wd-icon-check-small');
+      });
+    });
+
+    describe('with a value and background', () => {
+      test('should render a ColorInput with a value and the value as a background', () => {
```
Comment (anicholls) [inline modules/color-picker/react/spec/ColorInput.spec.tsx]: This could be combined with the value test on line 32.

File: modules/color-picker/react/spec/ColorInput.spec.tsx (hunk 7 above)
Comment (anicholls) [reply modules/color-picker/react/spec/ColorInput.spec.tsx]: Agreed! 👍 I think it's good to check that `onChange` and `onValidColorChange` is called, but a lot of the specific behavioral functionality (hash stripping, too long of input, etc.) could/should be checked in cypress

File: modules/form-field/react/stories/stories_ColorInput.tsx [hunk 11]
```diff
@@ -124,3 +150,61 @@ storiesOf('Components|Inputs/Color Picker/Color Input/React/Left Label', module)
       {controlComponent(<ColorInput />)}
     </FormField>
   ));
+
+storiesOf('Components|Inputs/Color Picker/Color Input/React/Visual Testing', module)
+  .addParameters({component: ColorInput})
+  .addDecorator(withReadme(README))
+  .add('States', () => (
+    <StaticStates>
```
Comment (anicholls) [inline modules/form-field/react/stories/stories_ColorInput.tsx]: Do you mind updating this with the changes in #418?

File: modules/color-picker/react/spec/ColorInput.spec.tsx (hunk 9 above)
Comment (sahlhoff) [reply modules/color-picker/react/spec/ColorInput.spec.tsx]: Will do!

File: modules/color-picker/react/spec/ColorInput.spec.tsx (hunk 10 above)
Comment (sahlhoff) [reply modules/color-picker/react/spec/ColorInput.spec.tsx]: Updated.

File: modules/form-field/react/stories/stories_ColorInput.tsx (hunk 11 above)
Comment (sahlhoff) [reply modules/form-field/react/stories/stories_ColorInput.tsx]: Updated.
